### PR TITLE
doc: cephfs: improve documentation of "ceph nfs cluster create" and "ceph fs volume create" commands

### PR DESCRIPTION
**Pull request by smithfarm**
When a reader sees "ceph nfs create <type=cephfs>" it's not
clear that this means they should type "ceph nfs create cephfs".

I also took this opportunity to clarify the command description
based on my testing and discussions with the CephFS developers.

Fixes: https://tracker.ceph.com/issues/46559
Signed-off-by: Nathan Cutler <ncutler@suse.com>

### CODE REVIEW

**Comment by smithfarm** [comment]
I'm hoping to use this PR to make further improvements to this documentation page. So far, I have been able to create an "NFS Ganesha cluster" and an export, and I could even mount the export successfully, but when I try to `touch` a file inside the mount it tells me the file system is read-only. It would be great if we could add an explanation of that part to this page.

One thing I noticed was that this doc page says that "ceph nfs cluster create" . . .

    . . . creates a common recovery pool for all Ganesha daemons, new user based on cluster_id

Now, when I do "ceph nfs export ls ... --detailed" I can see

    "user_id": "sesdevnfs1",

inside the `"fsal"` section of the export, but - as far as I can tell - no user was created.

Can anyone tell me what is missing so I could add it to this doc page?

**Comment by smithfarm** [comment]
Another thing that would be good to emphasize is that the user is not expected to use `ceph orch apply` to deploy NFS Ganesha. Does `ceph nfs cluster create` support the `-i` option, i.e. can I write a service specification in a YAML file and pass it to `ceph nfs cluster create` like I can to `ceph orch apply`? (I guess not, but in either case we should document it explicitly.)

**Comment by batrick** [comment]
@smithfarm it doesn't sound like you did anything wrong. Can you share the sequence of `ceph nfs` commands you ran?

**Comment by batrick** [comment]
> Another thing that would be good to emphasize is that the user is not expected to use `ceph orch apply` to deploy NFS Ganesha.

Right. Can you add that to this PR?

> Does `ceph nfs cluster create` support the `-i` option, i.e. can I write a service specification in a YAML file and pass it to `ceph nfs cluster create` like I can to `ceph orch apply`? (I guess not, but in either case we should document it explicitly.)

For what purpose?

**Comment by smithfarm** [comment]
> > Does `ceph nfs cluster create` support the `-i` option, i.e. can I write a service specification in a YAML file and pass it to `ceph nfs cluster create` like I can to `ceph orch apply`? (I guess not, but in either case we should document it explicitly.)
> 
> For what purpose?

My only purpose here is to increase the likelihood that following the documentation will lead to a useful result.

The orchestrator placement doc that is linked to here ("For more details on placement specification refer to the `orchestrator doc`") explains that there are two ways to specify the placement: as a string and as a yaml file. Now, as a user I find this confusing because the workflow for deploying NFS differs substantially from the workflow for deploying CephFS, and from the workflow for deploying other orchestrator services. With CephFS we are expected to first create the MDS service and start the daemon containers ("ceph orch apply mds" or "ceph orch apply" with a service spec file including a `service_type: mds` stanza) and then run `ceph fs volume create <volumename>` to deploy a CephFS over those daemons. But with NFS we are expected to only run `ceph nfs cluster create` and specify the service/daemon placement in an argument to that command.

**Comment by smithfarm** [comment]
> @smithfarm it doesn't sound like you did anything wrong. Can you share the sequence of `ceph nfs` commands you ran?

```
master:~ # ceph fs status
sesdevfs - 0 clients
========
RANK  STATE            MDS               ACTIVITY     DNS    INOS  
 0    active  sesdevfs.master.xykihv  Reqs:    0 /s    10     13   
        POOL            TYPE     USED  AVAIL  
cephfs.sesdevfs.meta  metadata  1536k  9000M  
cephfs.sesdevfs.data    data       0   9000M  
MDS version: ceph version 15.2.4-337-g55cec95eaf (55cec95eaf49c566262f3ac00c3dbf432c134f12) octopus (stable)

master:~ # ceph nfs cluster create cephfs sesdevnfs master

master:~ # ceph nfs cluster ls
sesdevnfs

master:~ # ceph nfs export create cephfs sesdevfs sesdevnfs nfs

master:~ # ceph nfs export ls sesdevnfs --detailed                                  
[
  {
    "export_id": 1,
    "path": "/",
    "cluster_id": "sesdevnfs",
    "pseudo": "nfs",
    "access_type": "RW",
    "squash": "no_root_squash",
    "security_label": true,
    "protocols": [
      4
    ],
    "transports": [
      "TCP"
    ],
    "fsal": {
      "name": "CEPH",
      "user_id": "sesdevnfs1",
      "fs_name": "sesdevfs",
      "sec_label_xattr": ""
    },
    "clients": []
  }
]

master:~ # mount -t nfs4 master:/ /mnt
master:~ # echo $?
0

master:~ # mount -l | grep ^master
master:/ on /mnt type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1)

master:~ # touch /mnt/bubba
touch: cannot touch '/mnt/bubba': Read-only file system
```

After that, I thought: maybe I need to create the "pseudo" (target NFS export path, `/nfs`) inside the CephFS. So I did that and tried:

```
master:~ # mount -t nfs4 master:/nfs /mnt
```

That just yielded a different error:

```
mount.nfs4: requested NFS version or transport protocol is not supported
```

Maybe this is because the `"fsal"` stanza of the export configuration created by `ceph nfs export create` does not include a secret?

I'd also like to add to the doc a hint how to display the log where one could further investigate why I'm getting this error "mount.nfs4: requested NFS version or transport protocol is not supported". 

I tried:

```
cephadm logs --name nfs.ganesha-sesdevnfs.master -- -n 100 >log.out
```

but no new log lines appear when I run the mount command, so I doubt that the error is causing NFS Ganesha to log anything.

**Comment by smithfarm** [comment]
Another suspicious thing I see in the NFS Ganesha log is this:

```
Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] sigmgr_thread :MAIN :EVENT :SIGHUP_HANDLER: Received SIGHUP.... initiating export list reload
Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] reclaim_reset :FSAL :EVENT :start_reclaim failed: -2
Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] export_commit_common :CONFIG :CRIT :A Pseudo path must be an absolute path
Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File ("rados://nfs-ganesha/sesdevnfs/export-1":1): 1 validation errors in block EXPORT
Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File ("rados://nfs-ganesha/sesdevnfs/export-1":1): Errors processing block (EXPORT)
```

Now, the export configuration created by `ceph nfs export create` specifies `pseudo` as a relative path, but NFS Ganesha complains: "A Pseudo path must be an absolute path".

And maybe this is the crux of the problem... the container I am using contains NFS-Ganesha version 4dev.22, not version 3.3 that `ceph nfs export create` is expecting.

**Comment by batrick** [comment]
> > > Does `ceph nfs cluster create` support the `-i` option, i.e. can I write a service specification in a YAML file and pass it to `ceph nfs cluster create` like I can to `ceph orch apply`? (I guess not, but in either case we should document it explicitly.)
> > 
> > 
> > For what purpose?
> 
> My only purpose here is to increase the likelihood that following the documentation will lead to a useful result.
> 
> The orchestrator placement doc that is linked to here ("For more details on placement specification refer to the `orchestrator doc`") explains that there are two ways to specify the placement: as a string and as a yaml file. Now, as a user I find this confusing because the workflow for deploying NFS differs substantially from the workflow for deploying CephFS, and from the workflow for deploying other orchestrator services.

That's reasonable. So you suggest there should also be a `ceph nfs cluster create -i` command. @varshar16 any objections?

> With CephFS we are expected to first create the MDS service and start the daemon containers ("ceph orch apply mds" or "ceph orch apply" with a service spec file including a `service_type: mds` stanza) and then run `ceph fs volume create <volumename>` to deploy a CephFS over those daemons. But with NFS we are expected to only run `ceph nfs cluster create`, specifying the service/daemon placement in an argument to that command.

I don't think this is right. The `fs volume` command is supposed to create the MDS daemons for you. Also, the mds_autoscaler (#32731) will also automate growing/shrinking the FS in the future.

**Comment by batrick** [comment]
> > @smithfarm it doesn't sound like you did anything wrong. Can you share the sequence of `ceph nfs` commands you ran?
> 
> ```
> master:~ # ceph fs status
> sesdevfs - 0 clients
> ========
> RANK  STATE            MDS               ACTIVITY     DNS    INOS  
>  0    active  sesdevfs.master.xykihv  Reqs:    0 /s    10     13   
>         POOL            TYPE     USED  AVAIL  
> cephfs.sesdevfs.meta  metadata  1536k  9000M  
> cephfs.sesdevfs.data    data       0   9000M  
> MDS version: ceph version 15.2.4-337-g55cec95eaf (55cec95eaf49c566262f3ac00c3dbf432c134f12) octopus (stable)
> 
> master:~ # ceph nfs cluster create cephfs sesdevnfs master
> 
> master:~ # ceph nfs cluster ls
> sesdevnfs
> 
> master:~ # ceph nfs export create cephfs sesdevfs sesdevnfs nfs
>
> master:~ # ceph nfs export ls sesdevnfs --detailed                                  
> [
>   {
>     "export_id": 1,
>     "path": "/",
>     "cluster_id": "sesdevnfs",
>     "pseudo": "nfs",
>     "access_type": "RW",
>     "squash": "no_root_squash",
>     "security_label": true,
>     "protocols": [
>       4
>     ],
>     "transports": [
>       "TCP"
>     ],
>     "fsal": {
>       "name": "CEPH",
>       "user_id": "sesdevnfs1",
>       "fs_name": "sesdevfs",
>       "sec_label_xattr": ""
>     },
>     "clients": []
>   }
> ]
> 
> master:~ # mount -t nfs4 master:/ /mnt
> master:~ # echo $?
> 0
> 
> master:~ # mount -l | grep ^master
> master:/ on /mnt type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1)
> 
> master:~ # touch /mnt/bubba
> touch: cannot touch '/mnt/bubba': Read-only file system
> ```

You're manipulating the virtual VFS ganesha creates. The actual mount is in `master:/nfs`.

> After that, I thought: maybe I need to create the "pseudo" (target NFS export path, `/nfs`) inside the CephFS. So I did that and tried:
> 
> ```
> master:~ # mount -t nfs4 master:/nfs /mnt
> ```
> 
> That just yielded a different error:
> 
> ```
> mount.nfs4: requested NFS version or transport protocol is not supported
> ```

That's odd. Suggest trying `-o nfsvers=4.1,proto=tcp`. Otherwise, maybe @jtlayton has an idea.

> Maybe this is because the `"fsal"` stanza of the export configuration created by `ceph nfs export create` does not include a secret?

No, that's not it.

> I'd also like to add to the doc a hint how to display the log where one could further investigate why I'm getting this error "mount.nfs4: requested NFS version or transport protocol is not supported".
> 
> I tried:
> 
> ```
> cephadm logs --name nfs.ganesha-sesdevnfs.master -- -n 100 >log.out
> ```
> 
> but no new log lines appear when I run the mount command, so I doubt that the error is causing NFS Ganesha to log anything.

NFS-Ganesha doesn't produce any debug information probably with this default deployments. We're looking at improving this by providing a custom config, see https://tracker.ceph.com/issues/45747

**Comment by batrick** [comment]
> Another suspicious thing I see in the NFS Ganesha log is this:
> 
> ```
> Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] sigmgr_thread :MAIN :EVENT :SIGHUP_HANDLER: Received SIGHUP.... initiating export list reload
> Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] reclaim_reset :FSAL :EVENT :start_reclaim failed: -2
> Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] export_commit_common :CONFIG :CRIT :A Pseudo path must be an absolute path
> Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File ("rados://nfs-ganesha/sesdevnfs/export-1":1): 1 validation errors in block EXPORT
> Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File ("rados://nfs-ganesha/sesdevnfs/export-1":1): Errors processing block (EXPORT)
> ```
> 
> Now, the export configuration created by `ceph nfs export create` specifies `pseudo` as a relative path, but NFS Ganesha complains: "A Pseudo path must be an absolute path".

Ah, that might be it. If that's the issue let's raise a tracker ticket so the nfs module catches this problem.

> And maybe this is the crux of the problem... the container I am using contains NFS-Ganesha version 4dev.22, not version 3.3 that `ceph nfs export create` is expecting.

**Comment by varshar16** [comment]
> > Another suspicious thing I see in the NFS Ganesha log is this:
> > ```
> > Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] sigmgr_thread :MAIN :EVENT :SIGHUP_HANDLER: Received SIGHUP.... initiating export list reload
> > Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] reclaim_reset :FSAL :EVENT :start_reclaim failed: -2
> > Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] export_commit_common :CONFIG :CRIT :A Pseudo path must be an absolute path
> > Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File ("rados://nfs-ganesha/sesdevnfs/export-1":1): 1 validation errors in block EXPORT
> > Jul 15 19:54:59 master bash[27312]: 15/07/2020 17:54:59 : epoch 5f0f41ef : master : ganesha.nfsd-1[sigmgr] config_errs_to_log :CONFIG :CRIT :Config File ("rados://nfs-ganesha/sesdevnfs/export-1":1): Errors processing block (EXPORT)
> > ```
> > 
> > 
> > Now, the export configuration created by `ceph nfs export create` specifies `pseudo` as a relative path, but NFS Ganesha complains: "A Pseudo path must be an absolute path".
> 
> Ah, that might be it. If that's the issue let's raise a tracker ticket so the nfs module catches this problem.
> 
Yes, it is the issue here. I have created a tracker ticket for it: https://tracker.ceph.com/issues/46565.

> > And maybe this is the crux of the problem... the container I am using contains NFS-Ganesha version 4dev.22, not version 3.3 that `ceph nfs export create` is expecting.

**Comment by varshar16** [comment]
> > > @smithfarm it doesn't sound like you did anything wrong. Can you share the sequence of `ceph nfs` commands you ran?
> > 
> > 
> > ```
> > master:~ # ceph fs status
> > sesdevfs - 0 clients
> > ========
> > RANK  STATE            MDS               ACTIVITY     DNS    INOS  
> >  0    active  sesdevfs.master.xykihv  Reqs:    0 /s    10     13   
> >         POOL            TYPE     USED  AVAIL  
> > cephfs.sesdevfs.meta  metadata  1536k  9000M  
> > cephfs.sesdevfs.data    data       0   9000M  
> > MDS version: ceph version 15.2.4-337-g55cec95eaf (55cec95eaf49c566262f3ac00c3dbf432c134f12) octopus (stable)
> > 
> > master:~ # ceph nfs cluster create cephfs sesdevnfs master
> > 
> > master:~ # ceph nfs cluster ls
> > sesdevnfs
> > 
> > master:~ # ceph nfs export create cephfs sesdevfs sesdevnfs nfs
> > 
> > master:~ # ceph nfs export ls sesdevnfs --detailed                                  
> > [
> >   {
> >     "export_id": 1,
> >     "path": "/",
> >     "cluster_id": "sesdevnfs",
> >     "pseudo": "nfs",
> >     "access_type": "RW",
> >     "squash": "no_root_squash",
> >     "security_label": true,
> >     "protocols": [
> >       4
> >     ],
> >     "transports": [
> >       "TCP"
> >     ],
> >     "fsal": {
> >       "name": "CEPH",
> >       "user_id": "sesdevnfs1",
> >       "fs_name": "sesdevfs",
> >       "sec_label_xattr": ""
> >     },
> >     "clients": []
> >   }
> > ]
> > 
> > master:~ # mount -t nfs4 master:/ /mnt
> > master:~ # echo $?
> > 0
> > 
> > master:~ # mount -l | grep ^master
> > master:/ on /mnt type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1)
> > 
> > master:~ # touch /mnt/bubba
> > touch: cannot touch '/mnt/bubba': Read-only file system
> > ```
> 
> You're manipulating the virtual VFS ganesha creates. The actual mount is in `master:/nfs`.
> 
> > After that, I thought: maybe I need to create the "pseudo" (target NFS export path, `/nfs`) inside the CephFS. So I did that and tried:
> > ```
> > master:~ # mount -t nfs4 master:/nfs /mnt
> > ```
> > 
> > 
> > That just yielded a different error:
> > ```
> > mount.nfs4: requested NFS version or transport protocol is not supported
> > ```
> 
> That's odd. Suggest trying `-o nfsvers=4.1,proto=tcp`. Otherwise, maybe @jtlayton has an idea.
> 
And I got this error with relative pseudo root path
```
[root@varsha build]# mount -t nfs -o port=2049 192.168.0.138:/cephfs /mnt
Created symlink /run/systemd/system/remote-fs.target.wants/rpc-statd.service → /usr/lib/systemd/system/rpc-statd.service.
mount.nfs: mounting 192.168.0.138:/cephfs failed, reason given by server: No such file or directory
```
@smithfarm  Can you try again with absolute pseudo root path ?

> > Maybe this is because the `"fsal"` stanza of the export configuration created by `ceph nfs export create` does not include a secret?
> 
> No, that's not it.
> 
> > I'd also like to add to the doc a hint how to display the log where one could further investigate why I'm getting this error "mount.nfs4: requested NFS version or transport protocol is not supported".
> > I tried:
> > ```
> > cephadm logs --name nfs.ganesha-sesdevnfs.master -- -n 100 >log.out
> > ```
> > 
> > 
> > but no new log lines appear when I run the mount command, so I doubt that the error is causing NFS Ganesha to log anything.
> 
> NFS-Ganesha doesn't produce any debug information probably with this default deployments. We're looking at improving this by providing a custom config, see https://tracker.ceph.com/issues/45747

**Comment by varshar16** [comment]
> > > > Does `ceph nfs cluster create` support the `-i` option, i.e. can I write a service specification in a YAML file and pass it to `ceph nfs cluster create` like I can to `ceph orch apply`? (I guess not, but in either case we should document it explicitly.)
> > > 
> > > 
> > > For what purpose?
> > 
> > 
> > My only purpose here is to increase the likelihood that following the documentation will lead to a useful result.
> > The orchestrator placement doc that is linked to here ("For more details on placement specification refer to the `orchestrator doc`") explains that there are two ways to specify the placement: as a string and as a yaml file. Now, as a user I find this confusing because the workflow for deploying NFS differs substantially from the workflow for deploying CephFS, and from the workflow for deploying other orchestrator services.
> 
> That's reasonable. So you suggest there should also be a `ceph nfs cluster create -i` command. @varshar16 any objections?
> 

We can do it.

> > With CephFS we are expected to first create the MDS service and start the daemon containers ("ceph orch apply mds" or "ceph orch apply" with a service spec file including a `service_type: mds` stanza) and then run `ceph fs volume create <volumename>` to deploy a CephFS over those daemons. But with NFS we are expected to only run `ceph nfs cluster create`, specifying the service/daemon placement in an argument to that command.
> 
> I don't think this is right. The `fs volume` command is supposed to create the MDS daemons for you. Also, the mds_autoscaler (#32731) will also automate growing/shrinking the FS in the future.

File: doc/cephfs/fs-nfs-exports.rst [hunk 1]
```diff
@@ -17,15 +17,30 @@ Create NFS Ganesha Cluster
 
 .. code:: bash
 
-    $ ceph nfs cluster create <type=cephfs> <clusterid> [<placement>]
+    $ ceph nfs cluster create cephfs <clusterid> [<placement>]
```
Comment (varshar16) [inline doc/cephfs/fs-nfs-exports.rst:20]: We intend to use the same interface for rgw. So don't remove the type.
```suggestion
    $ ceph nfs cluster create <type> <clusterid> [<placement>]
```

**Comment by smithfarm** [comment]
>  Can you try again with absolute pseudo root path ?

@varshar16 Yes, that did the trick - thank you! I went ahead and added a commit so this point is clear in the documentation.

@batrick I added a commit updating the "ceph fs volume create" documentation so it contains the same explanation of the placement string.

I'm not sure whether it makes sense to support placement YAML files or not. The idea behind the YAML files is that you can create one big YAML file specifying the placement of all the daemons in the cluster and then say:

    ceph orch apply -i cluster.yml

Also, the the orchestrator knows how to dump the current cluster state in one big YAML file, so if you need to change the number of daemons or which hosts they run on, you can dump the file, edit it, and then re-apply it using the same command.

@sebastian-philipp @jschmid1 @mgfritch Do you think it makes sense for "ceph fs volume create" and "ceph nfs cluster create" to support placement YAML files as well as placement strings? (Currently, they only support placement strings.)

File: doc/cephfs/fs-nfs-exports.rst (hunk 1 above)
Comment (smithfarm) [reply doc/cephfs/fs-nfs-exports.rst]: done - please take another look

File: doc/cephfs/fs-nfs-exports.rst [hunk 2]
```diff
@@ -17,15 +17,37 @@ Create NFS Ganesha Cluster
...
 .. code:: bash
 
-    $ ceph nfs cluster create <type=cephfs> <clusterid> [<placement>]
+    $ ceph nfs cluster create <type> <clusterid> [<placement>]
 
 This creates a common recovery pool for all Ganesha daemons, new user based on
-cluster_id and common ganesha config rados object.
+cluster_id, and a common ganesha config rados object. It can also bring up NFS
+Ganesha daemons using the enabled ceph-mgr orchestrator module (see
+:doc:`mgr/orchestrator`), e.g. cephadm.
```
Comment (smithfarm) [inline doc/cephfs/fs-nfs-exports.rst:25]: /home/jenkins-build/build/workspace/ceph-pr-docs/doc/cephfs/fs-nfs-exports.rst:22: WARNING: unknown document: mgr/orchestrator

**Comment by varshar16** [comment]
jenkins render docs

File: doc/cephfs/fs-nfs-exports.rst [hunk 3]
```diff
@@ -70,8 +92,14 @@ Create CephFS Export
 
     $ ceph nfs export create cephfs <fsname> <clusterid> <binding> [--readonly] [--path=/path/in/cephfs]
 
-It creates export rados objects containing the export block. Here binding is
-the pseudo root name and type is export type.
+This creates export rados objects containing the export block.
```
Comment (batrick) [inline doc/cephfs/fs-nfs-exports.rst:95]:
```suggestion
This creates export RADOS objects containing the export block.
```

File: doc/cephfs/fs-volumes.rst [hunk 4]
```diff
@@ -45,9 +45,29 @@ Create a volume using::
...
-This creates a CephFS file system and its data and metadata pools. It also tries
-to create MDSes for the filesystem using the enabled ceph-mgr orchestrator
-module  (see :doc:`/mgr/orchestrator`) , e.g., rook.
+This creates a CephFS file system and its data and metadata pools. It can also
+try to create MDSes for the filesystem using the enabled ceph-mgr orchestrator
+module (see :doc:`/mgr/orchestrator`), e.g. rook.
+
+<vol_name> is the volume name (an arbitrary string), and
+
+<placement> is an optional string signifying which hosts should have NFS Ganesha
```
Comment (sebastian-philipp) [inline doc/cephfs/fs-volumes.rst:54]: can we simply link to :ref:`orchestrator-cli-placement-spec` here?

Comment (smithfarm) [reply doc/cephfs/fs-volumes.rst:54]: @sebastian-philipp We need a place that documents **only** placement strings and doesn't mention YAML files. But @varshar16 says YAML file support is coming for `ceph fs` and `ceph fs`. Once that is implemented, then this section could be replaced with your link?

**Comment by smithfarm** [comment]
@batrick @varshar16 Is specifying `/` for the pseudo root path supposed to work? I found that it did not work, but like the relative path, "ceph nfs export create" did not complain about it.

**Comment by batrick** [comment]
> @batrick @varshar16 Is specifying `/` for the pseudo root path supposed to work? I found that it did not work, but like the relative path, "ceph nfs export create" did not complain about it.

Ya, it can't just be "/". @varshar16 another thing to catch.

File: doc/cephfs/fs-volumes.rst (hunk 4 above)
Comment (varshar16) [reply doc/cephfs/fs-volumes.rst:54]: Not for `ceph fs`, only for `ceph fs nfs` we can have YAML file support. The nfs yaml will not be same as `orch` yaml file and this yaml file will need to be transformed according to `orch` yaml file.

Example:
fs nfs yaml file
```
cluster type:
cluster id:
placement:
config:
```

Only the placement part has to be similar to `orch` yaml file.

**Comment by varshar16** [comment]
> > @batrick @varshar16 Is specifying `/` for the pseudo root path supposed to work? I found that it did not work, but like the relative path, "ceph nfs export create" did not complain about it.
> 
> Ya, it can't just be "/". @varshar16 another thing to catch.

@batrick @smithfarm I have updated this PR https://github.com/ceph/ceph/pull/36136

**Comment by smithfarm** [comment]
@batrick @varshar16 Have your change requests been addressed?

**Comment by batrick** [comment]
> @batrick @varshar16 Have your change requests been addressed?

Think you missed: https://github.com/ceph/ceph/pull/36125#discussion_r455886042

**Comment by batrick** [comment]
> > @batrick @varshar16 Have your change requests been addressed?
> 
> Think you missed: [#36125 (comment)](https://github.com/ceph/ceph/pull/36125#discussion_r455886042)

@smithfarm ping

**Comment by smithfarm** [comment]
@batrick Thanks for the reminder: rebased and added two more commits. (In other words, only the stuff in the last two commits is new.)

File: doc/cephfs/fs-nfs-exports.rst [hunk 5]
```diff
@@ -17,15 +17,37 @@ Create NFS Ganesha Cluster
 
 .. code:: bash
 
-    $ ceph nfs cluster create <type=cephfs> <clusterid> [<placement>]
+    $ ceph nfs cluster create <type> <clusterid> [<placement>]
 
-This creates a common recovery pool for all Ganesha daemons, new user based on
-cluster_id and common ganesha config rados object.
+This creates a common recovery pool for all NFS Ganesha daemons, new user based on
+cluster_id, and a common NFS Ganesha config RADOS object. It can also bring up
```
Comment (varshar16) [inline doc/cephfs/fs-nfs-exports.rst:23]: It uses the orch module to create daemons
```suggestion
cluster_id, and a common NFS Ganesha config RADOS object. It brings up
```

**Comment by varshar16** [comment]
jenkins render docs

File: doc/cephfs/fs-nfs-exports.rst [hunk 6]
```diff
@@ -89,8 +113,14 @@ Create CephFS Export
 
     $ ceph nfs export create cephfs <fsname> <clusterid> <binding> [--readonly] [--path=/path/in/cephfs]
 
-It creates export rados objects containing the export block. Here binding is
-the pseudo root name and type is export type.
+This creates export RADOS objects containing the export block.
+
+<fsname> is the name of the FS volume used by the NFS Ganesha cluster that will
```
Comment (varshar16) [inline doc/cephfs/fs-nfs-exports.rst:118]:
```suggestion
``fsname`` is the name of the FS volume used by the NFS Ganesha cluster that will
```

**Review by varshar16 — APPROVED** [review]
The angular braces look weird for defining the arguments. Otherwise looks good.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test api

**Comment by smithfarm** [comment]
@batrick This PR only touches stuff under `doc/` and has passed all the mandatory tests. Do we have to get the non-mandatory tests to pass as well?

**Comment by batrick** [comment]
> @batrick This PR only touches stuff under `doc/` and has passed all the mandatory tests. Do we have to get the non-mandatory tests to pass as well?

api tests are about to be made mandatory. I realize this PR can't feasibly break the api tests.

**Pull request merged**